java-spiffe-core/src/main/java/io/spiffe/internal/CertificateUtils.java

```diff
@@ -0,0 +1,269 @@
+     *
+     * @param chain a list of {@link X509Certificate}
+     * @return a {@link TrustDomain}
+     * @throws CertificateException
+     */
+    public static TrustDomain getTrustDomain(final List<X509Certificate> chain) throws CertificateException {
+        val spiffeId = getSpiffeId(chain.get(0));
+        return spiffeId.getTrustDomain();
+    }
+
+    /**
+     * Validates that the private key and the public key in the x509Certificate match by
+     * creating a signature with the private key and verifying with the public key.
+     *
+     * @throws InvalidKeyException if the keys don't match
+     */
+    public static void validatePrivateKey(final PrivateKey privateKey, final X509Certificate x509Certificate) throws InvalidKeyException {
+        Algorithm.Family algorithm = Algorithm.Family.parse(privateKey.getAlgorithm());
+        if (Algorithm.Family.RSA.equals(algorithm)) {
+            verifyKeys(privateKey, x509Certificate.getPublicKey(), SHA_512_WITH_RSA);
+        } else if (Algorithm.Family.EC.equals(algorithm)) {
+            verifyKeys(privateKey, x509Certificate.getPublicKey(), SHA_512_WITH_ECDSA);
+        } else {
+            throw new InvalidKeyException(String.format("Private Key algorithm not supported: %s", algorithm));
+        }
+    }
+
+    public static boolean isCA(final X509Certificate cert) {
+        return cert.getBasicConstraints() != -1;
+    }
+
+    public static boolean hasKeyUsageCertSign(final X509Certificate cert) {
+        boolean[] keyUsage = cert.getKeyUsage();
+        return keyUsage[KEY_CERT_SIGN];
+    }
+
+    public static boolean hasKeyUsageDigitalSignature(final X509Certificate cert) {
+        boolean[] keyUsage = cert.getKeyUsage();
+        return keyUsage[DIGITAL_SIGNATURE];
+    }
+
+    public static boolean hasKeyUsageCRLSign(final X509Certificate cert) {
+        boolean[] keyUsage = cert.getKeyUsage();
+        return keyUsage[CRL_SIGN];
+    }
+
+    private static void verifyKeys(final PrivateKey privateKey, final PublicKey publicKey, final String algorithm) throws InvalidKeyException {
+        final String randomString = RandomStringUtils.random(100, 0, 0, true, true, null, new SecureRandom());
+        byte[] challenge = randomString.getBytes();
+
+        try {
+            Signature sig = Signature.getInstance(algorithm);
+            sig.initSign(privateKey);
+            sig.update(challenge);
+            byte[] signature = sig.sign();
+            sig.initVerify(publicKey);
+            sig.update(challenge);
+            if (!sig.verify(signature)) {
+                throw new InvalidKeyException("Private Key does not match Certificate Public Key");
+            }
+        } catch (NoSuchAlgorithmException | SignatureException e) {
+            throw new InvalidKeyException("Private and Public Keys could not be verified", e);
+        }
+    }
+
+    private static List<String> getSpiffeIds(final X509Certificate certificate) throws CertificateParsingException {
```

Comment:
Well, the method is private and returning a List is more useful because I need to `get` an element from the returned collection, I will keep as it is.